nexus/reconfigurator/planning/src/blueprint_builder/builder/datasets_editor.rs

```diff
@@ -0,0 +1,358 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Helper for editing the datasets of a Blueprint
+
+use crate::planner::PlannerRng;
+
+use super::Ensure;
+use illumos_utils::zpool::ZpoolName;
+use nexus_types::deployment::BlueprintDatasetConfig;
+use nexus_types::deployment::BlueprintDatasetDisposition;
+use nexus_types::deployment::BlueprintDatasetsConfig;
+use nexus_types::deployment::SledResources;
+use nexus_types::deployment::ZpoolFilter;
+use omicron_common::api::external::ByteCount;
+use omicron_common::api::external::Generation;
+use omicron_common::disk::CompressionAlgorithm;
+use omicron_common::disk::DatasetKind;
+use omicron_common::disk::DatasetName;
+use omicron_common::disk::GzipLevel;
+use omicron_uuid_kinds::DatasetUuid;
+use omicron_uuid_kinds::SledUuid;
+use omicron_uuid_kinds::ZpoolUuid;
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::net::SocketAddrV6;
+
+#[derive(Debug, thiserror::Error)]
+pub enum BlueprintDatasetsEditError {
+    #[error(
+        "internal inconsistency: multiple datasets with kind {kind:?} \
+         on zpool {zpool_id}: {id1}, {id2}"
+    )]
+    MultipleDatasetsOfKind {
+        zpool_id: ZpoolUuid,
+        kind: DatasetKind,
+        id1: DatasetUuid,
+        id2: DatasetUuid,
+    },
+}
+
+/// Helper for working with sets of datasets on each sled
+///
+/// Tracking the set of datasets is slightly non-trivial because we need to
+/// bump the per-sled generation number iff the datasets are changed.  So
+/// we need to keep track of whether we've changed the datasets relative
+/// to the parent blueprint.
+#[derive(Debug)]
+pub(super) struct BlueprintDatasetsEditor {
+    current: BTreeMap<SledUuid, BlueprintDatasetsConfig>,
+    changed: BTreeSet<SledUuid>,
+}
+
+impl BlueprintDatasetsEditor {
+    pub fn new(current: BTreeMap<SledUuid, BlueprintDatasetsConfig>) -> Self {
+        Self { current, changed: BTreeSet::new() }
+    }
+
+    /// Get a helper to edit the datasets of a specific sled.
+    ///
+    /// If any changes are made via the returned editor, the sled will be
+    /// recorded as needing a generation bump in its dataset config when the
+    /// editor is dropped.
+    pub fn sled_datasets_editor<'a>(
+        &'a mut self,
+        sled_id: SledUuid,
+        sled_resources: &SledResources,
+        rng: &'a mut PlannerRng,
+    ) -> Result<SledDatasetsEditor<'a>, BlueprintDatasetsEditError> {
+        let config = self
+            .current
+            .entry(sled_id)
+            .or_insert_with(empty_blueprint_datasets_config);
+
+        // Gather all dataset IDs known to the blueprint and to the database.
+        let blueprint_dataset_ids =
+            build_dataset_kind_id_map(config.datasets.values().map(
+                |dataset| (dataset.pool.id(), dataset.kind.clone(), dataset.id),
+            ))?;
```

Comment:
`blueprint_dataset_ids` is derived from `config`, right?

Can't we just pass `config` to `SledDatasetsEditor::new`, and have the callee do the reconstruction of the set of `blueprint_dataset_ids`? Otherwise, it seems possible to pass a config / blueprint_dataset_ids pair that are incompatible.

Reply:
Yes, good call; fixed in [085594d](https://github.com/oxidecomputer/omicron/pull/7104/commits/085594d3a4974b15292e0dc6d62bed35a6e9d188)